application/models/SurveyDynamic.php

```diff
@@ -518,12 +524,9 @@ public function timeline($sType, $dStart, $dEnd)
     /**
      * @return bool
      */
-    private function getbHaveToken()
+    public function getbHaveToken()
     {
-        if (!isset($this->bHaveToken)) {
-            $this->bHaveToken = tableExists('tokens_'.self::$sid) && Permission::model()->hasSurveyPermission(self::$sid, 'tokens', 'read'); // Boolean : show (or not) the token;
-        }
-        return $this->bHaveToken;
+        return tableExists('tokens_'.self::$sid) && Permission::model()->hasSurveyPermission(self::$sid, 'tokens', 'read'); // Boolean : show (or not) the token;
```

Comment:
A general comment. I do not think its a good idea to make permission checks in models. This should be covered in other places (eg controller).

Reply:
General comment : some model function NEED Permission : you think we can add a array var 

```
public var $permission = [
'create' => false,
'read' => false,
'update' => false,
'delete' => false,
'import' => false,
'export' => false,
]
```

At minima for buttons it's needed (in my opinion).

Maybe you can open a Development mantis about this ? If we are OK , i can start on it in develop version.